## Common preparations

### Importing necessary libs.

In [2]:
import pandas as pd
import os

### Defining paths to data files.

In [3]:
base_dir = os.path.join('data', 'bookmaker')
log_filename = os.path.join(base_dir, 'log.csv')
users_filename = os.path.join(base_dir, 'users.csv')

### Loading log and users data from CSV files into corresponding Pandas DataFrames and merging them.

In [4]:
#Функция, редактирующую значения столбца win
def fillna_win(row):
    if row.bet==0: #если ставка bet=0, то выигрыш win=0
        return 0
    elif row.bet>0 and row.win>0: #если ставка bet>0 и выигрыш win>0, то выигрыш win=win
        return row.win
    else:
        return -row.bet #если ставка bet>0 и выигрыш win=0, то выигрыш win=-bet
#Функция, формирующая значения столбца net
def column_net(row):
    if row.win<0: #если выигрыш win<0, то net=win
        return row.win
    else: #если выигрыш win>=0, то net=win-bet
        return row.win-row.bet
    
#Готовим данные
#log
log = pd.read_csv(log_filename, header=None)
log.columns = ['user_id','time','bet','win']

log.bet = log.bet.fillna(0)
log.win = log.apply(lambda row: fillna_win(row), axis=1)
log = log[log.user_id != '#error']
log.user_id = log.user_id.str.split(' - ').apply(lambda x: x[1])
log['net'] = log.apply(lambda row: column_net(row), axis=1)
log['time'] = log['time'].apply(lambda x: x[1:])
log.time = pd.to_datetime(log['time'])

#users
us = pd.read_csv(users_filename,sep='\t', encoding='koi8_r')
us.columns = ['user_id','mail','city']
us.user_id = us.user_id.apply(lambda x: x.lower())

#Слияние
log_users = pd.merge(log, us, on='user_id')

## Tasks.

### Task 1.
Используйте датасет, который получился в результате всех преобразований в прошлой секции (в том числе, заполнение пропусков).

Ответ запишите в поле ниже в виде одного слова, с большой буквы.
Наибольший суммарный выигрыш среди всех городов имеет Москва. Посчитайте следующий за ней город. Для этого сгруппируйте по городам, возьмите признак win, просуммируйте по каждому городу, отсортируйте и получите второй город.

### Solution:

In [17]:
log_users.groupby('city')['win'].sum().sort_values(ascending=False).index[1]

'Воронеж'

### Conclusion: 'Воронеж' is the right answer!!!

### Task 2.
Используйте датасет, который получился в результате всех преобразований в прошлой секции (в том числе, заполнение пропусков).

Ответ запишите в поле ниже в виде целого числа (нужно отбросить дробную часть).

Подсказка: можно использовать методы min() и max().
Во сколько раз различаются в среднем максимальная и минимальная ставки по городам? Для того, чтобы это посчитать, нужно сгруппировать по городам, взять среднее от признака bet, найти максимальное и минимальное значения, затем поделить одно на другое.

### Solution:

In [16]:
mean_min_bet = log_users.groupby('city')['bet'].mean().min()
mean_max_bet = log_users.groupby('city')['bet'].mean().max()
print(mean_max_bet/mean_min_bet)

227.3904449648712


### Conclusion: 227 will be the right answer!!!

### Task 3.


Давайте повторим все, что мы прошли в этой секции. Напишите код, который посчитает, сколько раз пользователи приходили в букмекерскую контору в каждом городе. Для этого:

1. загрузите датасеты log.csv и users.csv;

2. удалите user_id с ошибкой (#error) и приведите признак user_id к одному виду в обоих датасетах;

3. слейте два датасета в один по признаку user_id;

4. сгруппируйте данные по правильному признаку (какому - вам нужно понять самим), затем выберите user_id и воспользуйтесь функцией count() для подсчета наблюдений в каждой группе;

5. результат (таблицу) запишите в sample2.

### Solution:

In [19]:
#Функция, редактирующую значения столбца win
def fillna_win(row):
    if row.bet==0: #если ставка bet=0, то выигрыш win=0
        return 0
    elif row.bet>0 and row.win>0: #если ставка bet>0 и выигрыш win>0, то выигрыш win=win
        return row.win
    else:
        return -row.bet #если ставка bet>0 и выигрыш win=0, то выигрыш win=-bet
#Функция, формирующая значения столбца net
def column_net(row):
    if row.win<0: #если выигрыш win<0, то net=win
        return row.win
    else: #если выигрыш win>=0, то net=win-bet
        return row.win-row.bet
    
#Готовим данные
#log
log = pd.read_csv(log_filename, header=None)
log.columns = ['user_id','time','bet','win']

log.bet = log.bet.fillna(0)
log.win = log.apply(lambda row: fillna_win(row), axis=1)
log = log[log.user_id != '#error']
log.user_id = log.user_id.str.split(' - ').apply(lambda x: x[1])
log['net'] = log.apply(lambda row: column_net(row), axis=1)
log['time'] = log['time'].apply(lambda x: x[1:])
log.time = pd.to_datetime(log['time'])

#users
us = pd.read_csv(users_filename,sep='\t', encoding='koi8_r')
us.columns = ['user_id','mail','city']
us.user_id = us.user_id.apply(lambda x: x.lower())

#Слияние
log_users = pd.merge(log, us, on='user_id')

sample2 = log_users.groupby('city')['user_id'].count()

### Conclusion: the code above is one of the right answers!!!